### ISSUE TITLE
Support complex joiner networks

### ISSUE THREAD

**Issue by csukuangfj** [awni/transducer#8]
I find that the current implementation supports only joiner networks containing an adder:
https://github.com/awni/transducer/blob/b517f1f60177b6be2e3928a11c02784de8977672/torch_test.py#L187-L188

In the paper [Speech Recognition with Deep Recurrent Neural Networks](https://arxiv.org/pdf/1303.5778.pdf), which is a successor of [Sequence Transduction with Recurrent Neural Networks](https://arxiv.org/abs/1211.3711), it mentions another kind of joiner network:

> In the original formulation Pr(k|t, u) was defined by taking an ‘acoustic’ distribution Pr(k|t) from the CTC network,
a ‘linguistic’ distribution Pr(k|u) from the prediction network, then multiplying the two together and renormalising.
An improvement introduced in this paper is to instead feed
the hidden activations of both networks into a separate feedforward output network, whose outputs are then normalised
with a softmax function to yield Pr(k|t, u). This allows a
richer set of possibilities for combining linguistic and acoustic information, and appears to lead to better generalisation.
In particular we have found that the number of deletion errors
encountered during decoding is reduced.

Wondering whether it will support feedfoward joiner networks that contain `nn.Linear()` and `nn.Tanh()` layers.

---

Also, the README.md says:

> The memory of this implementation scales with the product B * T * U and does not increase with the token set size

But the memory occupied by the encoder and decoder is proportional to the vocabulary size. The memory consumed by a vocab size of 10k is certainly more than that of a size of 1k, I believe.

**Comment by awni** [comment]
Yes, I added a comment to the README. The trade off of this implementation is that you cannot use a joiner network, it hard codes the additive combination of emissions and predictions.

Regarding memory use, you are correct that the encoder and deecoder memory use scales with vocab size (B * T *V) and (B * U * V) respectively. But the RNNT implementation scales as (B * T * U) which is the point of this code.

**Comment by csukuangfj** [comment]
> The trade off of this implementation is that you cannot use a joiner network, it hard codes the additive combination of emissions and predictions

Could you share some performance results, e.g., in ASR training, about using such a simple joiner ?
